contracts/contracts/Coordinator.sol

```diff
@@ -0,0 +1,123 @@
+    event ConfirmationPosted(uint32 indexed ritualID, address indexed node, address[] confirmedNodes);
+
+    enum RitualState {
+        NON_INITIATED,
+        WAITING_FOR_TRANSCRIPTS,
+        WAITING_FOR_CONFIRMATIONS,
+        ENDED
+    }
+
+    // TODO: Find better name
+    struct Performance {
+        address node;
+        uint96 confirmedBy;
+        bytes32 transcript;
+    }
+
+    struct Ritual {
+        uint32 id;
+        uint32 initTimestamp;
+        Performance[DKG_SIZE] performance;
+    }
+
+    uint32 public immutable transcriptsWindow;
+    uint32 public immutable confirmationsWindow;
+    Ritual[] public rituals;
+
+    constructor(uint32 _transcriptsWindow, uint32 _confirmationsWindow){
+        transcriptsWindow = _transcriptsWindow;
+        confirmationsWindow = _confirmationsWindow;
+    }
+
+    function numberOfRituals() external view returns(uint256){
+        return rituals.length;
+    }
+
+    function initiateRitual(address[] calldata nodes) external {
+        // TODO: Check for payment
+        // TODO: Check for expiration time
+        // TODO: Improve DKG size choices
+        require(nodes.length == DKG_SIZE, "Invalid number of nodes");
+
+
+        uint32 id = uint32(rituals.length);
+        Ritual storage ritual = rituals.push();
+        ritual.id = id;
+        ritual.initTimestamp = uint32(block.timestamp);
+
+        address previousNode = nodes[0];
+        ritual.performance[0].node = previousNode;
+        address currentNode;
+        for(uint256 i=1; i < nodes.length; i++){
+            currentNode = nodes[i];
+            require(currentNode > previousNode, "Nodes must be sorted");
+            ritual.performance[i].node = currentNode;
+            previousNode = currentNode;
+            // TODO: Check nodes are eligible (staking, etc)
+        }
+
+        emit NewRitual(id, nodes);
+    }
+
+    function postTranscript(uint32 ritualID, uint256 nodeIndex, bytes calldata transcript) external {
+        Ritual storage ritual = rituals[ritualID];
+        require(getRitualState(ritual) == RitualState.WAITING_FOR_TRANSCRIPTS, "Not waiting for transcripts");
+        require(ritual.performance[nodeIndex].node == msg.sender, "Node not part of ritual");
+
+        // Nodes commit to their transcript
+        bytes32 transcriptDigest = keccak256(transcript);
+        ritual.performance[nodeIndex].transcript = transcriptDigest;
+        
+        emit TranscriptPosted(ritualID, msg.sender, transcriptDigest);
+    }
+
+    function postConfirmation(uint32 ritualID, uint256 nodeIndex, uint256[] calldata confirmedNodesIndexes) external {
+        Ritual storage ritual = rituals[ritualID];
+        require(getRitualState(ritual) == RitualState.WAITING_FOR_CONFIRMATIONS, "Not waiting for confirmations");
+        require(ritual.performance[nodeIndex].transcript != bytes32(0), "Node not part of ritual");
```

Comment:
do we need to ensure that `msg.sender` is:
- not the node that is being confirmed
- within the ritual

Reply:
Right, good catch.

Reply:
Fixed in 1a7aab7d3f4c1f459188df6c66b9f1a5383165ac. Note that msg.sender can confirm itself explicitly if they want, but the contract already does it when they confirm others, so it's just a waste of gas.